src/Validator/UniqueUrlValidator.php

```diff
@@ -36,16 +39,20 @@ public function __construct(PageManagerInterface $manager)
 
     public function validate($value, Constraint $constraint): void
     {
-        if (!$value instanceof PageInterface) {
-            $this->context->addViolation('The page is not valid, expected a PageInterface');
-
+        if (!$constraint instanceof UniqueUrl) {
+            throw new UnexpectedTypeException($constraint, UniqueUrl::class);
+        }
+        // do not validate on null, use NotNull instead
+        if (null === $value) {
             return;
         }
+        if (!$value instanceof PageInterface) {
+            throw new UnexpectedValueException($value, PageInterface::class);
+        }
 
+        // at this it can only be SiteInterface or null
         if (!$value->getSite() instanceof SiteInterface) {
-            $this->context->addViolation('The page is not linked to a Site');
-
-            return;
+            throw new UnexpectedValueException($value->getSite(), SiteInterface::class);
```

Comment:
Shouldn't it be a addViolation with a message ?

Reply:
i'm following the route of other Validators there, for example the Language Validator throws this when it can't be turned into a String

https://github.com/symfony/symfony/blob/870eeb975feb1abb4b8a1722e1fd57beeab2b230/src/Symfony/Component/Validator/Constraints/LanguageValidator.php#L40-L42

Reply:
I agree it should be done for `if (!$value instanceof PageInterface) {` because it means that the constraint is used on a wrong value.

But for the check `if (!$value->getSite() instanceof SiteInterface) {` I think it's better to have a message violation to say the user didn't selected a Site. (Like we had before)

Reply:
for the Page -> getSite thing i wasn't sure when directly getSite returns something else than SiteInterface
imo that can only happen for null right?

Reply:
otherwise if it is a message, it should be translated like `error.uniq_url` is?

Reply:
https://github.com/sonata-project/SonataPageBundle/blob/3.x/src/Model/PageInterface.php#L330-L332

So yes, either getSite is a site, either it's null.
And when it's null, it would mean that the user should select a site. (which is supposed to be always true https://github.com/sonata-project/SonataPageBundle/blob/3.x/src/Admin/PageAdmin.php#L298)

Reply:
> otherwise if it is a message, it should be translated like `error.uniq_url` is?

Yes, using a translation key would be waaaaaay better indeed :)

Reply:
I added better translatable Violation message